usr.sbin/amd/amd/amd.8
.\" $NetBSD: amd.8,v 1.12 1999/02/01 19:05:09 christos Exp $
.\"
.\"
.\" Copyright (c) 1997-1999 Erez Zadok
.\" Copyright (c) 1989 Jan-Simon Pendry
.\" Copyright (c) 1989 Imperial College of Science, Technology & Medicine
.\" Copyright (c) 1989 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Jan-Simon Pendry at Imperial College, London.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"      This product includes software developed by the University of
.\"      California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	%W% (Berkeley) %G%
.\"
.\" Id: amd.8,v 1.2 1999/01/10 21:53:40 ezk Exp
.\"
.TH AMD 8 "3 November 1989"
.SH NAME
amd \- automatically mount file systems
.SH SYNOPSIS
.B amd
.B \-H
.br
.B amd
[
.BI \-F " conf_file"
]
.br
.B amd
[
.B \-nprvHS
] [
.BI \-a " mount_point"
] [
.BI \-c " duration"
] [
.BI \-d " domain"
] [
.BI \-k " kernel-arch"
] [
.BI \-l " logfile"
] [
.BI \-o " op_sys_ver"
] [
.BI \-t " interval.interval"
] [
.BI \-w " interval"
] [
.BI \-x " log-option"
] [
.BI \-y " YP-domain"
] [
.BI \-C " cluster-name"
] [
.BI \-D " option"
] [
.BI \-F " conf_file"
] [
.BI \-O " op_sys_name"
] [
.BI \-T " tag"
]
[
.I directory
.I mapname
.RI [ " \-map-options " ]
] .\|.\|.
.SH DESCRIPTION
.B Amd
is a daemon that automatically mounts filesystems
whenever a file or directory
within that filesystem is accessed.
Filesystems are automatically unmounted when they
appear to have become quiescent.
.LP
.B Amd
operates by attaching itself as an
.SM NFS
server to each of the specified
.IB directories .
Lookups within the specified directories
are handled by
.BR amd ,
which uses the map defined by
.I mapname
to determine how to resolve the lookup.
Generally, this will be a host name, some filesystem information
and some mount options for the given filesystem.
.LP
In the first form depicted above,
.B amd
will print a short help string.  In the second form, if no options are
specified, or the
.B -F
is used,
.B amd
will read configuration parameters from the file
.I conf_file
which defaults to
.BR /etc/amd.conf .
The last form is described below.
.SH OPTIONS

.\"*******************************************************"

.TP
.BI \-a " temporary-directory"
Specify an alternative location for the real mount points.
The default is
.BR /a .

.TP
.BI \-c " duration"
Specify a
.IR duration ,
in seconds, that a looked up name remains
cached when not in use.  The default is 5 minutes.

.TP
.BI \-d " domain"
Specify the local domain name.  If this option is not
given the domain name is determined from the hostname.

.TP
.BI \-k " kernel-arch"
Specifies the kernel architecture.  This is used solely
to set the ${karch} selector.

.TP
.BI \-l " logfile"
Specify a logfile in which to record mount and unmount events.
If
.I logfile
is the string
.B syslog
then the log messages will be sent to the system log daemon by
.IR syslog (3).
The default syslog facility used is LOG_DAEMON.  If you
wish to change it, append its name to the log file name, delimited by a
single colon.  For example, if
.I logfile
is the string
.B syslog:local7
then
.B Amd
will log messages via
.IR syslog (3)
using the LOG_LOCAL7 facility (if it exists on the system).

.TP
.B \-n
Normalize hostnames.
The name refereed to by ${rhost} is normalized relative to the
host database before being used.  The effect is to translate
aliases into ``official'' names.

.TP
.BI \-o " op_sys_ver"
Override the compiled-in version number of the operating system.  Useful
when the built in version is not desired for backward compatibility reasons.
For example, if the build in version is ``2.5.1'', you can override it to
``5.5.1'', and use older maps that were written with the latter in mind.

.TP
.B \-p
Print PID.
Outputs the process-id of
.B amd
to standard output where it can be saved into a file.

.TP
.B \-r
Restart existing mounts.
.B Amd
will scan the mount file table to determine which filesystems
are currently mounted.  Whenever one of these would have
been auto-mounted,
.B amd
.I inherits
it.

.TP
.BI \-t " interval.interval"
Specify the
.IR interval ,
in tenths of a second, between NFS/RPC/UDP retries.
The default is 0.8 seconds.
The second values alters the restransmit counter.
Useful defaults are supplied if either or both
values are missing.

.TP
.B \-v
Version.  Displays version and configuration information on standard error.

.TP
.BI \-w " interval"
Specify an
.IR interval ,
in seconds, between attempts to dismount
filesystems that have exceeded their cached times.
The default is 2 minutes.

.TP
.BI \-x " options"
Specify run-time logging options.  The options are a comma separated
list chosen from: fatal, error, user, warn, info, map, stats, all.

.TP
.BI \-y " domain"
Specify an alternative NIS domain from which to fetch the NIS maps.
The default is the system domain name.  This option is ignored if NIS
support is not available.

.TP
.BI \-C " cluster-name"
Specify an alternative HP-UX cluster name to use.

.TP
.BI \-D " option"
Select from a variety of debug options.  Prefixing an
option with the strings
.B no
reverses the effect of that option.  Options are cumulative.
The most useful option is
.BR all .
Since
.I \-D
is only used for debugging other options are not documented here:
the current supported set of options is listed by the \-v option
and a fuller description is available in the program source.

.TP
.BI \-F " conf_file"
Specify an amd configuration file to use.  See
.BR amd.conf (5)
for description of this file's format.  This configuration file is used to
specify any options in lieu of typing many of them on the command line.  The
.I amd.conf
file includes directives for every command line option amd has, and many
more that are only available via the configuration file facility.  The
configuration file specified by this option is processed after all other
options had been processed, regardless of the actual location of this option
on the command line.

.TP
.B \-H
Print help and usage string.

.TP
.BI \-O " op_sys_name"
Override the compiled-in name of the operating system.  Useful when the
built in name is not desired for backward compatibility reasons.  For
example, if the build in name is ``sunos5'', you can override it to
``sos5'', and use older maps which were written with the latter in mind.

.TP
.B \-S
Do not lock the running executable pages of amd into memory.  To improve
amd's performance, systems that support the
.BR plock (3)
call, could lock the amd process into memory.  This way there is less chance
the operating system will schedule, page out, and swap the amd process as
needed.  This tends improves amd's performance, at the cost of reserving the
memory used by the amd process (making it unavailable for other processes).
If this behavior is not desired, use the
.B \-S
option.

.TP
.BI \-T " tag"
Specify a tag to use with
.BR amd.conf (5).
All map entries tagged with
.I tag
will be processed.  Map entries that are not tagged are always processed.
Map entries that are tagged with a tag other than
.I tag
will not be processed.

.SH FILES
.PD 0
.TP 5
.B /a
directory under which filesystems are dynamically mounted
.TP 5
.B /etc/amd.conf
default configuration file
.PD
.SH CAVEATS
Some care may be required when creating a mount map.
.LP
Symbolic links on an NFS filesystem can be incredibly inefficient.
In most implementations of NFS, their interpolations are not cached
by the kernel and each time a symlink is encountered during a
.I lookuppn
translation it costs an RPC call to the NFS server.
It would appear that a large improvement in real-time
performance could be gained by adding a cache somewhere.
Replacing symlinks with a suitable incarnation of the auto-mounter
results in a large real-time speedup, but also causes a large
number of process context switches.
.LP
A weird imagination is most useful to gain full advantage of all
the features.
.SH "SEE ALSO"
.BR amd.conf (5),
.BR amq (8),
.BR domainname (1),
.BR hostname (1),
.BR automount (8),
.BR mount (8),
.BR umount (8),
.BR mtab (5),
.BR syslog (3).
.LP
.I "Amd \- The 4.4 BSD Automounter"
.SH AUTHORS
Jan-Simon Pendry <jsp@doc.ic.ac.uk>, Department of Computing, Imperial College, London, UK.
.P
Erez Zadok <ezk@cs.columbia.edu>, Department of Computer Science, Columbia
University, New York, USA.
.P
Other authors and contributors to am-utils are listed in the
.B AUTHORS
file distributed with am-utils.


usr.sbin/amd/amd/amd.c
/*	$NetBSD: amd.c,v 1.12 1999/09/04 22:38:17 christos Exp $	*/

/*
 * Copyright (c) 1997-1999 Erez Zadok
 * Copyright (c) 1989 Jan-Simon Pendry
 * Copyright (c) 1989 Imperial College of Science, Technology & Medicine
 * Copyright (c) 1989 The Regents of the University of California.
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Jan-Simon Pendry at Imperial College, London.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgment:
 *      This product includes software developed by the University of
 *      California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *      %W% (Berkeley) %G%
 *
 * Id: amd.c,v 1.5 1999/02/04 07:24:14 ezk Exp
 *
 */

/*
 * Automounter
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif /* HAVE_CONFIG_H */
#include <am_defs.h>
#include <amd.h>

struct amu_global_options gopt;	/* where global options are stored */

char pid_fsname[16 + MAXHOSTNAMELEN];	/* "kiska.southseas.nz:(pid%d)" */
char *hostdomain = "unknown.domain";
char hostd[2 * MAXHOSTNAMELEN + 1]; /* Host+domain */
char *endian = ARCH_ENDIAN;	/* Big or Little endian */
char *cpu = HOST_CPU;		/* CPU type */
char *PrimNetName;		/* name of primary network */
char *PrimNetNum;		/* number of primary network */

int immediate_abort;		/* Should close-down unmounts be retried */
int orig_umask = 022;
int select_intr_valid;

jmp_buf select_intr;
struct amd_stats amd_stats;	/* Server statistics */
struct in_addr myipaddr;	/* (An) IP address of this host */
time_t do_mapc_reload = 0;	/* mapc_reload() call required? */

#ifdef HAVE_SIGACTION
sigset_t masked_sigs;
#endif /* HAVE_SIGACTION */


/*
 * Signal handler:
 * SIGINT - tells amd to do a full shutdown, including unmounting all
 *       filesystem.
 * SIGTERM - tells amd to shutdown now.  Just unmounts the automount nodes.
 */
static RETSIGTYPE
sigterm(int sig)
{
#ifdef REINSTALL_SIGNAL_HANDLER
  signal(sig, sigterm);
#endif /* REINSTALL_SIGNAL_HANDLER */

  switch (sig) {
  case SIGINT:
    immediate_abort = 15;
    break;

  case SIGTERM:
    immediate_abort = -1;
    /* fall through... */

  default:
    plog(XLOG_WARNING, "WARNING: automounter going down on signal %d", sig);
    break;
  }
  if (select_intr_valid)
    longjmp(select_intr, sig);
}


/*
 * Hook for cache reload.
 * When a SIGHUP arrives it schedules a call to mapc_reload
 */
static RETSIGTYPE
sighup(int sig)
{
#ifdef REINSTALL_SIGNAL_HANDLER
  signal(sig, sighup);
#endif /* REINSTALL_SIGNAL_HANDLER */

#ifdef DEBUG
  if (sig != SIGHUP)
    dlog("spurious call to sighup");
#endif /* DEBUG */
  /*
   * Force a reload by zero'ing the timer
   */
  if (amd_state == Run)
    do_mapc_reload = 0;
}


static RETSIGTYPE
parent_exit(int sig)
{
  exit(0);
}


static int
daemon_mode(void)
{
  int bgpid;

#ifdef HAVE_SIGACTION
  struct sigaction sa, osa;

  sa.sa_handler = parent_exit;
  sa.sa_flags = 0;
  sigemptyset(&(sa.sa_mask));
  sigaddset(&(sa.sa_mask), SIGQUIT);
  sigaction(SIGQUIT, &sa, &osa);
#else /* not HAVE_SIGACTION */
  signal(SIGQUIT, parent_exit);
#endif /* not HAVE_SIGACTION */

  bgpid = background();

  if (bgpid != 0) {
    /*
     * Now wait for the automount points to
     * complete.
     */
    for (;;)
      pause();
    /* should never reach here */
  }
#ifdef HAVE_SIGACTION
  sigaction(SIGQUIT, &osa, NULL);
#else /* not HAVE_SIGACTION */
  signal(SIGQUIT, SIG_DFL);
#endif /* not HAVE_SIGACTION */

  /*
   * Record our pid to make it easier to kill the correct amd.
   */
  if (gopt.flags & CFM_PRINT_PID) {
    if (STREQ(gopt.pid_file, "/dev/stdout")) {
      printf("%ld\n", (long) am_mypid);
      fflush(stdout);
      /* do not fclose stdout */
    } else {
      FILE *f;
      mode_t prev_umask = umask(0022); /* set secure temporary umask */

      f = fopen(gopt.pid_file, "w");
      if (f) {
	fprintf(f, "%ld\n", (long) am_mypid);
	(void) fclose(f);
      } else {
	fprintf(stderr, "cannot open %s (errno=%d)\n", gopt.pid_file, errno);
      }
      umask(prev_umask);	/* restore umask */
    }
  }

  /*
   * Pretend we are in the foreground again
   */
  foreground = 1;

  /*
   * Dissociate from the controlling terminal
   */
  amu_release_controlling_tty();

  return getppid();
}


/*
 * Initialize global options structure.
 */
static void
init_global_options(void)
{
#if defined(HAVE_SYS_UTSNAME_H) && defined(HAVE_UNAME)
  static struct utsname un;
#endif /* defined(HAVE_SYS_UTSNAME_H) && defined(HAVE_UNAME) */

  memset(&gopt, 0, sizeof(struct amu_global_options));

  /* name of current architecture */
  gopt.arch = HOST_ARCH;

  /* automounter temp dir */
  gopt.auto_dir = "/a";

  /* cluster name */
  gopt.cluster = NULL;

  /*
   * kernel architecture: this you must get from uname() if possible.
   */
#if defined(HAVE_SYS_UTSNAME_H) && defined(HAVE_UNAME)
  if (uname(&un) >= 0)
    gopt.karch = un.machine;
  else
#endif /* defined(HAVE_SYS_UTSNAME_H) && defined(HAVE_UNAME) */
    gopt.karch = HOST_ARCH;

  /* amd log file */
  gopt.logfile = NULL;

  /* operating system name */
  gopt.op_sys = HOST_OS_NAME;

  /* OS version */
  gopt.op_sys_ver = HOST_OS_VERSION;

  /* pid file */
  gopt.pid_file = "/dev/stdout";

  /* local domain */
  gopt.sub_domain = NULL;

  /* NFS retransmit counter */
  gopt.amfs_auto_retrans = -1;

  /* NFS retry interval */
  gopt.amfs_auto_timeo = -1;

  /* cache duration */
  gopt.am_timeo = AM_TTL;

  /* dismount interval */
  gopt.am_timeo_w = AM_TTL_W;

  /*
   * various CFM_* flags.
   * by default, only the "plock" option is on (if available).
   */
  gopt.flags = CFM_PROCESS_LOCK;

#ifdef HAVE_MAP_HESIOD
  /* Hesiod rhs zone */
  gopt.hesiod_base = "automount";
#endif /* HAVE_MAP_HESIOD */

#ifdef HAVE_MAP_LDAP
  /* LDAP base */
  gopt.ldap_base = NULL;

  /* LDAP host ports */
  gopt.ldap_hostports = NULL;

  /* LDAP cache */
  gopt.ldap_cache_seconds = 0;
  gopt.ldap_cache_maxmem = 131072;
#endif /* HAVE_MAP_LDAP */

#ifdef HAVE_MAP_NIS
  /* YP domain name */
  gopt.nis_domain = NULL;
#endif /* HAVE_MAP_NIS */
}


int
main(int argc, char *argv[])
{
  char *domdot, *verstr;
  int ppid = 0;
  int error;
  char *progname = NULL;		/* "amd" */
  char hostname[MAXHOSTNAMELEN + 1] = "localhost"; /* Hostname */
#ifdef HAVE_SIGACTION
  struct sigaction sa;
#endif /* HAVE_SIGACTION */

  /*
   * Make sure some built-in assumptions are true before we start
   */
  assert(sizeof(nfscookie) >= sizeof(u_int));
  assert(sizeof(int) >= 4);

  /*
   * Set processing status.
   */
  amd_state = Start;

  /*
   * Determine program name
   */
  if (argv[0]) {
    progname = strrchr(argv[0], '/');
    if (progname && progname[1])
      progname++;
    else
      progname = argv[0];
  }
  if (!progname)
    progname = "amd";
  am_set_progname(progname);

  /*
   * Initialize process id.  This is kept
   * cached since it is used for generating
   * and using file handles.
   */
  am_set_mypid();

  /*
   * Get local machine name
   */
  if (gethostname(hostname, sizeof(hostname)) < 0) {
    plog(XLOG_FATAL, "gethostname: %m");
    going_down(1);
  }
  hostname[sizeof(hostname) - 1] = '\0';

  /*
   * Check it makes sense
   */
  if (!*hostname) {
    plog(XLOG_FATAL, "host name is not set");
    going_down(1);
  }

#ifdef DEBUG
  /* initialize debugging flags (Register AMQ, Enter daemon mode) */
  debug_flags = D_AMQ | D_DAEMON;
#endif /* DEBUG */

  /*
   * Initialize global options structure.
   */
  init_global_options();

  /*
   * Partially initialize hostd[].  This
   * is completed in get_args().
   */
  if ((domdot = strchr(hostname, '.'))) {
    /*
     * Hostname already contains domainname.
     * Split out hostname and domainname
     * components
     */
    *domdot++ = '\0';
    hostdomain = domdot;
  }
  strcpy(hostd, hostname);
  am_set_hostname(hostname);

  /*
   * Trap interrupts for shutdowns.
   */
#ifdef HAVE_SIGACTION
  sa.sa_handler = sigterm;
  sa.sa_flags = 0;
  sigemptyset(&(sa.sa_mask));
  sigaddset(&(sa.sa_mask), SIGINT);
  sigaddset(&(sa.sa_mask), SIGTERM);
  sigaction(SIGINT, &sa, NULL);
  sigaction(SIGTERM, &sa, NULL);
#else /* not HAVE_SIGACTION */
  (void) signal(SIGINT, sigterm);
#endif /* not HAVE_SIGACTION */

  /*
   * Trap Terminate so that we can shutdown gracefully (some chance)
   */
#ifdef HAVE_SIGACTION
  sa.sa_handler = sigterm;
  sa.sa_flags = 0;
  sigemptyset(&(sa.sa_mask));
  sigaddset(&(sa.sa_mask), SIGTERM);
  sigaction(SIGTERM, &sa, NULL);
#else /* not HAVE_SIGACTION */
  (void) signal(SIGTERM, sigterm);
#endif /* not HAVE_SIGACTION */

  /*
   * Hangups tell us to reload the cache
   */
#ifdef HAVE_SIGACTION
  sa.sa_handler = sighup;
  sa.sa_flags = 0;
  sigemptyset(&(sa.sa_mask));
  sigaddset(&(sa.sa_mask), SIGHUP);
  sigaction(SIGHUP, &sa, NULL);
#else /* not HAVE_SIGACTION */
  (void) signal(SIGHUP, sighup);
#endif /* not HAVE_SIGACTION */

  /*
   * Trap Death-of-a-child.  These allow us to
   * pick up the exit status of backgrounded mounts.
   * See "sched.c".
   */
#ifdef HAVE_SIGACTION
  sa.sa_handler = sigchld;
  sa.sa_flags = 0;
  sigemptyset(&(sa.sa_mask));
  sigaddset(&(sa.sa_mask), SIGCHLD);
  sigaction(SIGCHLD, &sa, NULL);

  /*
   * construct global "masked_sigs" used in nfs_start.c
   */
  sigemptyset(&masked_sigs);
  sigaddset(&masked_sigs, SIGHUP);
  sigaddset(&masked_sigs, SIGCHLD);
  sigaddset(&masked_sigs, SIGTERM);
  sigaddset(&masked_sigs, SIGINT);
#else /* not HAVE_SIGACTION */
  (void) signal(SIGCHLD, sigchld);
#endif /* not HAVE_SIGACTION */

  /*
   * Fix-up any umask problems.  Most systems default
   * to 002 which is not too convenient for our purposes
   */
  orig_umask = umask(0);

  /*
   * Figure out primary network name
   */
  getwire(&PrimNetName, &PrimNetNum);

  /*
   * Determine command-line arguments
   */
  get_args(argc, argv);

  /*
   * Log version information.
   */
  verstr = strtok(get_version_string(), "\n");
  plog(XLOG_INFO, "AM-UTILS VERSION INFORMATION:");
  while (verstr) {
    plog(XLOG_INFO, verstr);
    verstr = strtok(NULL, "\n");
  }

  /*
   * Get our own IP address so that we
   * can mount the automounter.
   */
  amu_get_myaddress(&myipaddr);
  plog(XLOG_INFO, "My ip addr is %s", inet_ntoa(myipaddr));

  /* avoid hanging on other NFS servers if started elsewhere */
  if (chdir("/") < 0)
    plog(XLOG_INFO, "cannot chdir to /: %m");

  /*
   * Now check we are root.
   */
  if (geteuid() != 0) {
    plog(XLOG_FATAL, "Must be root to mount filesystems (euid = %ld)", (long) geteuid());
    going_down(1);
  }

  /*
   * Lock process text and data segment in memory.
   */
#ifdef HAVE_PLOCK
  if (gopt.flags & CFM_PROCESS_LOCK) {
    if (plock(PROCLOCK) != 0) {
      plog(XLOG_WARNING, "Couldn't lock process text and data segment in memory: %m");
    } else {
      plog(XLOG_INFO, "Locked process text and data segment in memory");
    }
  }
#endif /* HAVE_PLOCK */

#ifdef HAVE_MAP_NIS
  /*
   * If the domain was specified then bind it here
   * to circumvent any default bindings that may
   * be done in the C library.
   */
  if (gopt.nis_domain && yp_bind(gopt.nis_domain)) {
    plog(XLOG_FATAL, "Can't bind to NIS domain \"%s\"", gopt.nis_domain);
    going_down(1);
  }
#endif /* HAVE_MAP_NIS */

#ifdef DEBUG
  amuDebug(D_DAEMON)
#endif /* DEBUG */
    ppid = daemon_mode();

  sprintf(pid_fsname, "%s:(pid%ld)", am_get_hostname(), (long) am_mypid);

  do_mapc_reload = clocktime() + ONE_HOUR;

  /*
   * Register automounter with system.
   */
  error = mount_automounter(ppid);
  if (error && ppid)
    kill(ppid, SIGALRM);
  going_down(error);

  abort();
  return 1; /* should never get here */
}
